### PR TITLE
protocol state body in transaction snark

### PR DESCRIPTION
**Pull request by deepthiskumar**
Added protocol state body as a witness to generate transaction snark and plugged in`curr_global_slot` in transaction snark  to check for transaction expiry (Haven't implemented this for when applying transactions outside the snark in `transaction_logic.ml`, will have a follow up pr)

Checklist:

- [ ] Tests were added for the new behavior
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them:

Closes #0000
Closes #0000

### CODE REVIEW

File: src/lib/staged_ledger/staged_ledger.ml [hunk 1]
```diff
@@ -724,7 +731,8 @@ module T = struct
           (Float.of_int @@ List.length work) ;
         Gauge.set Scan_state_metrics.snark_work_required
           (Float.of_int
-             (List.length (Scan_state.all_work_pairs_exn t.scan_state))) )
+             (List.length (Scan_state.all_work_statements_exn t.scan_state)))
+    )
 
   let apply t witness ~logger ~verifier ~state_body_hash =
```
Comment (emberian) [inline src/lib/staged_ledger/staged_ledger.ml]: should this `state_body_hash` be renamed?

File: src/lib/coda_transition/external_transition.ml [hunk 2]
```diff
@@ -892,7 +892,8 @@ module Staged_ledger_validation = struct
       Staged_ledger.apply ~logger ~verifier parent_staged_ledger
         staged_ledger_diff
         ~state_body_hash:
-          Protocol_state.(Body.hash @@ body parent_protocol_state)
+          ( Protocol_state.hash parent_protocol_state
```
Comment (psteckler) [inline src/lib/coda_transition/external_transition.ml]: notice this is hashing the body twice

**Comment by mrmr1993** [comment]
I've adopted this PR. Pushed commits to
* address PR comments so far
* pipe the `curr_global_slot` to `transaction_logic.ml`
  - eventually this will be a `Coda_base.Transaction_protocol_state.t` when I integrate this with tokens support
* bring up-to-date with develop
* add a test for consistent behaviour for transactions from timed accounts

File: src/lib/bootstrap_controller/bootstrap_controller.ml [hunk 3]
```diff
@@ -651,6 +670,18 @@ let%test_module "Bootstrap_controller tests" =
...
+              let get_state hash =
+                match
+                  Transition_frontier.find_protocol_state frontier hash
+                with
+                | Some protocol_state ->
+                    Ok protocol_state
+                | None ->
+                    Or_error.errorf
+                      !"Protocol state (for scan state transactions) for \
+                        %{sexp:State_hash.t} not found when"
```
Comment (mrmr1993) [inline src/lib/bootstrap_controller/bootstrap_controller.ml:682]: Nit: remove `when`

File: src/lib/transition_frontier/transition_frontier.ml [hunk 4]
```diff
@@ -438,13 +438,21 @@ module For_tests = struct
         let genesis_ledger =
           Lazy.force (Precomputed_values.genesis_ledger precomputed_values)
         in
+        (*scan state is empty so no protocol state should be required*)
+        let get_state hash =
+          Or_error.errorf
+            !"Protocol state (for scan state transactions) for \
+              %{sexp:State_hash.t} not found when"
```
Comment (mrmr1993) [inline src/lib/transition_frontier/transition_frontier.ml:445]: Nit: remove `when`

File: genesis_ledgers/phase_three/annotated_ledger.json [hunk 5]
```diff
@@ -1,2998 +1,212 @@
...
-  },
-  {
-   "pk": "4vsRCVmwWh3SviHCowUvxRNM7qub9KWowwkcTLteBJYQLL3BCmnrHqYfTjrMYi6w6F88Te81nrVjuNtXTa84ppK62SCwNh1rVKwuqoaiJd881PaDWyBoBQc5G8hpXxzACixDHmnh8755v2zw",
-   "sk": null,
-   "balance": "0.000000000",
-   "delegate": null,
-   "discord_username": "CodaBP4"
-  }
- ]
+ {
```
Comment (mrmr1993) [inline genesis_ledgers/phase_three/annotated_ledger.json:2]: Not sure this should be removed as part of this PR

File: src/config/testnet_postake_medium_curves.mlh [hunk 6]
```diff
@@ -1,7 +1,7 @@
 [%%define ledger_depth 14]
 [%%import "/src/config/curve/medium.mlh"]
 [%%import "/src/config/coinbase/realistic.mlh"]
-[%%import "/src/config/scan_state/point2tps.mlh"]
+[%%import "/src/config/scan_state/medium.mlh"]
```
Comment (mrmr1993) [inline src/config/testnet_postake_medium_curves.mlh:4]: Are these changes meant to be part of this PR?

File: genesis_ledgers/phase_three/annotated_ledger.json (hunk 5 above)
Comment (deepthiskumar) [reply genesis_ledgers/phase_three/annotated_ledger.json:2]: These are not. I added them to run a small qa-net before landing this on develop. Will revert after validating the qa-net

File: src/config/testnet_postake_medium_curves.mlh (hunk 6 above)
Comment (deepthiskumar) [reply src/config/testnet_postake_medium_curves.mlh:4]: same

**Pull request merged**